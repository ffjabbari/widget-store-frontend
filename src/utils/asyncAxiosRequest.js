/** Asynchronous axios request function. */

import axios from 'axios';

const BASE_URL = process.env.REACT_APP_BASE_URL || 'http://localhost:3001';

/** asyncAxiosRequest()
 *
 * Function to abstract error handling of axios requests.
 *
 * @param {string} url
 * @param {string} method
 * @param {object} data
 * @param {object} params
 * @param {string} baseURL
 */

export const asyncAxiosRequest = async (
  url,
  method,
  data,
  params,
  baseURL = BASE_URL
) => {
  try {
    const resp = await axios({ url, method, data, params, baseURL });
    // console.log(resp);
    return resp;
  } catch (error) {
    return { error };
  }
};
