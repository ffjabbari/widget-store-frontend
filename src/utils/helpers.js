import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
