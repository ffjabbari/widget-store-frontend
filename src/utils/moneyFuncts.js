/** Functions for basic money calculations for widget-store. */

import Decimal from 'decimal.js';

/* SF sales tax. */
const SF_SALES_TAX = '0.085';

/** calculateDiscountPrice()
 *
 * Calculate discounted price from product object.
 * Product object should have properties "price"
 * and "discount".
 * @param {object} product
 *
 * returns: string
 */

export function calculateDiscountPrice(product) {
  if (product.discount === '0.00') return product.price;
  const price = new Decimal(product.price);
  return price.minus(price.times(product.discount)).toFixed(2);
}

/** calculateSubtotal()
 *
 * Calculate items subtotal from cart.items object.
 * Each cart item should have properties "price", "discount",
 * and "quantity" of items ordered.
 *
 * @param {object} items {product_id: item, product_id: item}
 *
 * returns: string
 */

export function calculateSubtotal(items) {
  return Object.values(items)
    .reduce((acc, product) => {
      const price = new Decimal(calculateDiscountPrice(product));
      const itemTotal = price.times(product.quantity);
      return itemTotal.plus(acc);
    }, 0)
    .toFixed(2);
}

/** calculateDiscountPercent()
 *
 * Calculate discount as an integer percent from
 * product object.discount.
 *
 * E.g. 0.15 --> 15
 *
 * returns: string
 */

export function calculateDiscountPercent(product) {
  const discount = new Decimal(product.discount);
  return discount.times(100).toFixed(0);
}

/** getDiscountData()
 *
 * Return discounted price and discount as a percent
 * from product obj.
 * @param {object} product
 *
 * returns { discountPrice, percent }
 */

export function getDiscountData(product) {
  return {
    discountPrice: calculateDiscountPrice(product),
    percent: calculateDiscountPercent(product),
  };
}

/** calculateTax()
 *
 * Calculate tax if customer is in California.
 * @param {string} subtotal
 * @param {string} state
 *
 * returns: string
 */

export function calculateTax(subtotal, state) {
  if (!/^ca(lifornia)?$/i.test(state)) return '0.00';
  subtotal = new Decimal(subtotal);
  return subtotal.times(SF_SALES_TAX).toFixed(2);
}

/** calculateTotal()
 *
 * Calculate subtotal plus tax, plus shipping cost if provided.
 * @param {string} subtotal
 * @param {string} state
 * @param {string||number} shippingCost
 *
 * returns: string
 */

export function calculateTotal(subtotal, state, shippingCost) {
  subtotal = new Decimal(subtotal);
  const tax = calculateTax(subtotal, state);
  shippingCost = shippingCost || 0;
  return subtotal.plus(tax).plus(shippingCost).toFixed(2);
}
