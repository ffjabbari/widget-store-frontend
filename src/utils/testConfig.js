/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'The only remote you will need',
    description: 'Lorem ipsum dolor sit, placeat.',
    discount: '0.15',
    image_url:
      'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    quantity: 1,
    rating: null,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: 'Apt 5',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 1,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };

  // mock return from /order/:id
  testData.orderGet = {
    customer: 1,
    customer_info: {
      first_name: 'Test',
      last_name: 'User',
      email: 'foo@gmail.com',
      address: '123 Main St',
      address_line2: '',
      city: 'Big City',
      state: 'TX',
      postal_code: '39488',
      phone_number: '(555) 555-5555',
    },

    discount: null,
    items: [
      { 0: { order_id: 2, product_id: 6, quantity: 1 } },
      { 1: { order_id: 2, product_id: 8, quantity: 1 } },
    ],

    order_date: '2021-02-05T08:00:00.000Z',
    order_id: 2,
    shipping_address: null,
    shipping_cost: '12.00',
    shipping_method: {
      details: { cost: '12.00', name: 'USPS ground shipping (3-7 days)' },
      shipping_method: 'usps_ground',
    },

    status: 'Pending',
    subtotal: '784.16',
    tax: '66.65',
    total: '862.81',
    total_items_quantity: 2,
    tracking_number: null,
  };

  TEST_DATA.shippingMethods = {
    usps_ground: {
      cost: '12.00',
      name: 'USPS ground shipping (3-7 days)',
    },
    usps_priority: {
      cost: '16.00',
      name: 'USPS priority shipping (2-4 days)',
    },
    usps_next_day: {
      cost: '25.00',
      name: 'USPS next day delivery (1 day)',
    },
    ups_ground: {
      cost: '16.00',
      name: 'UPS ground shipping (3-6 days)',
    },
    ups_2nd_day: {
      cost: '20.00',
      name: 'UPS second day delivery (2 days)',
    },
    ups_overnight: {
      cost: '28.00',
      name: 'UPS overnight delivery (1 day)',
    },
  };
};
