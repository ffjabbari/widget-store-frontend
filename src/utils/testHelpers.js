import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
