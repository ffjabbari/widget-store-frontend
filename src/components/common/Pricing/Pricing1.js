/** Utility component to show Product price.
 *
 * If there is a discount it will show original price
 * with a strikethrough and provide a sale byline
 * showing the percent off.
 */

import styled from 'styled-components';

// local imports
import { getDiscountData } from '../../../utils/moneyFuncts';

// styled components:

// strike through original price if product has been discounted.
const PriceStrikethrough = styled.span`
  text-decoration: ${props => (props.discount > 0 ? 'line-through' : 'none')};
  color: red;
`;
// make price dark unless struck-through
const Price = styled.span`
  color: ${props => (props.discount > 0 ? 'var(--secondary)' : 'var(--dark)')};
  font-weight: ${props => (props.discount > 0 ? 'normal' : 'bold')};
`;
// sale byline span
const SaleByline = styled.span`
  display: block;
  color: var(--success);
  font-weight: bold;
  font-style: italic;
`;
// discounted price span
const DiscountedPriceSpan = styled.span`
  color: var(--dark);
  font-weight: bold;
  font-style: italic;
  background: var(--warning);
  border-radius: 5px;
`;

const Pricing1 = ({ product }) => {
  // if there is a discount create sale byline and discounted price span
  if (product.discount > 0) {
    const { discountPrice, percent } = getDiscountData(product);

    var saleByline = (
      <SaleByline className="mb-1">Sale {percent}% OFF</SaleByline>
    );
    // create discounted price span
    var discountPriceSpan = (
      <DiscountedPriceSpan className="ml-2 p-1">
        ${discountPrice}
      </DiscountedPriceSpan>
    );
  }

  return (
    <>
      {saleByline}
      <PriceStrikethrough discount={product.discount}>
        <Price discount={product.discount} className="price">
          ${product.price}
        </Price>
      </PriceStrikethrough>
      {discountPriceSpan}
    </>
  );
};

export default Pricing1;
