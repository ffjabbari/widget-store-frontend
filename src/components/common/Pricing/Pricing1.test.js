/** Pricing1 tests */

// local imports
import { act } from '@testing-library/react';
import { renderWithStore } from '../../../utils/testHelpers';
import Pricing1 from './Pricing1';

let product;

beforeEach(() => {
  product = {
    byline: 'Latest generation smart televison 2021',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    rating: null,
  };
});

test('renders Pricing1', () => {
  act(() => {
    renderWithStore(<Pricing1 product={product} />);
  });
});

test('Pricing1 snapshot', async () => {
  let asFragment;
  await act(async () => {
    const resp = renderWithStore(<Pricing1 product={product} />);
    asFragment = resp.asFragment;
  });
  expect(asFragment()).toMatchSnapshot();
});

test('Pricing1 w/ discount snapshot', async () => {
  product.discount = 0.15;
  let asFragment;
  await act(async () => {
    const resp = renderWithStore(<Pricing1 product={product} />);
    asFragment = resp.asFragment;
  });
  expect(asFragment()).toMatchSnapshot();
});

test('Shows sale byline w/ discount', async () => {
  product.discount = 0.15;
  let getByText;
  await act(async () => {
    const resp = renderWithStore(<Pricing1 product={product} />);
    getByText = resp.getByText;
  });
  const saleByline = getByText(/Sale \d+% OFF/);
  expect(saleByline).toBeInTheDocument();
});

test("Doesn't shows sale byline if no discount", async () => {
  let queryByText;
  await act(async () => {
    const resp = renderWithStore(<Pricing1 product={product} />);
    queryByText = resp.queryByText;
  });
  const saleByline = queryByText(/Sale \d+% OFF/);
  expect(saleByline).not.toBeInTheDocument();
});
