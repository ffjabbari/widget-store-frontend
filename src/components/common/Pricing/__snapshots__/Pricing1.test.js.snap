// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pricing1 snapshot 1`] = `
<DocumentFragment>
  <span
    class="sc-bdfBwQ ehKgzr"
  >
    <span
      class="sc-gsTCUz cxQGHo price"
    >
      $400.40
    </span>
  </span>
</DocumentFragment>
`;

exports[`Pricing1 w/ discount snapshot 1`] = `
<DocumentFragment>
  <span
    class="sc-dlfnbm gHhlng mb-1"
  >
    Sale 15% OFF
  </span>
  <span
    class="sc-bdfBwQ bnWraN"
  >
    <span
      class="sc-gsTCUz jbCYhM price"
    >
      $400.40
    </span>
  </span>
  <span
    class="sc-hKgILt hnYLUB ml-2 p-1"
  >
    $340.34
  </span>
</DocumentFragment>
`;
